lib/heroku_rails_deploy/deployer.rb

```diff
@@ -0,0 +1,103 @@
+require 'optparse'
+require 'yaml'
+
+module HerokuRailsDeploy
+  class Deployer
+    attr_reader :config_file, :args
+
+    class Options < Struct.new(:environment, :revision)
+      def self.create_default(app_registry)
+        new(app_registry.keys.first, 'HEAD')
+      end
+    end
+
+    def initialize(config_file, args)
+      @config_file = config_file
+      @args = args
+    end
+
+    def run
+      raise "Missing config file #{config_file}" unless File.file?(config_file)
+      app_registry = YAML.load(File.read(config_file))
+
+      options = Options.create_default(app_registry)
+      OptionParser.new do |parser|
+        parser.on_tail('-h', '--help', 'Show this message') do
+          puts parser
+          # rubocop:disable Lint/NonLocalExitFromIterator
+          return
+          # rubocop:enable Lint/NonLocalExitFromIterator
+        end
+
+        parser.on('-e', '--environment ENVIRONMENT',
+                  "The environment to deploy to. Must be in #{app_registry.keys.join(', ')} (default #{app_registry.keys.first})") do |environment|
+          options.environment = environment
+        end
+
+        parser.on('-r', '--revision REVISION',
+                  'The git revision to push. (default HEAD)') do |revision|
+          options.revision = revision
+        end
+      end.parse!(args)
+
+      app_name = app_registry.fetch(options.environment) do
+        raise OptionParser::InvalidArgument.new("Invalid environment '#{options.environment}'. " \
+          "Must be in #{app_registry.keys.join(', ')}")
+      end
+
+      puts "Pushing code to Heroku app #{app_name} for environment #{options.environment}"
+      push_code(app_name, options.revision)
+
+      puts 'Checking for pending migrations'
+      if pending_migrations?(app_name)
+        puts 'Running migrations'
+        run_migrations(app_name)
+        puts 'Restarting dynos'
+        restart_dynos(app_name)
+      else
+        puts 'No migrations required'
+      end
+    end
+
+    def push_code(app_name, revision)
+      run_command!("git push --force git@heroku.com:#{app_name}.git #{revision}:master")
```

Comment:
According to the Circle [docs](https://circleci.com/docs/continuous-deployment-with-heroku/#pre-or-post) we should pass in `$CIRCLE_SHA1` for the revision.